File: src/main/java/frc/robot/subsystems/coralshooter/CoralShooterSubsystem.java
package frc.robot.subsystems.coralshooter;

import static edu.wpi.first.units.Units.Amps;
import static edu.wpi.first.units.Units.Celsius;
import static edu.wpi.first.units.Units.RPM;

import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import frc.robot.subsystems.CS_SubsystemBase;
import frc.robot.subsystems.coralshooter.CoralShooterInterface.CoralShooterValues;
import frc.utils.CS_Utils;

public class CoralShooterSubsystem extends CS_SubsystemBase {
  private CoralShooterInterface coralShooterInterface;
  private CoralShooterValues values;
  private AngularVelocity shootingRPMLeft = RPM.of(0); // = CoralShooterConstants.RPMShootLeft;
  private AngularVelocity shootingRPMRight = RPM.of(0); // = CoralShooterConstants.RPMShootRight;

  public CoralShooterSubsystem(CoralShooterInterface subsystem_interface) {
    super();

    this.coralShooterInterface = subsystem_interface;
    values = new CoralShooterValues();
    println("Created");
  }

  public void startRampUp(AngularVelocity newRPMLeft, AngularVelocity newRPMRight) {
    shootingRPMLeft = newRPMLeft;
    shootingRPMRight = newRPMRight;
    coralShooterInterface.startShooter(shootingRPMLeft, shootingRPMRight);
  }

  public void setShooterRPM(AngularVelocity new_RPMLeft, AngularVelocity new_RPMRight) {
    shootingRPMLeft = new_RPMLeft;
    shootingRPMRight = new_RPMRight;
    coralShooterInterface.updateRPMShooter(new_RPMLeft, new_RPMRight);
  }

  public void startIntake() {
    coralShooterInterface.startShooter(
        CoralShooterConstants.RPMIntake, CoralShooterConstants.RPMIntake);
    coralShooterInterface.startLauncher(CoralShooterConstants.launcherIntakeSetpoint);
  }

  public void startLauncher(double new_Setpoint) {
    coralShooterInterface.startLauncher(new_Setpoint);
  }

  public void stopShooter() {
    coralShooterInterface.stopShooter();
  }

  public void stopLauncher() {
    coralShooterInterface.stopLauncher();
  }

  public void stopAll() {
    coralShooterInterface.stopShooter();
    coralShooterInterface.stopLauncher();
  }

  public AngularVelocity getShooterRPMLeft() {
    return coralShooterInterface.getRPMLeft();
  }

  public AngularVelocity getShooterRPMRight() {
    return coralShooterInterface.getRPMRight();
  }

  public boolean isLoaded() {
    return coralShooterInterface.isLoaded();
  }

  public boolean hasCoral() {
    return coralShooterInterface.hasCoral();
  }

  public void setPIDLeft(double newkP, double newkI, double newkD) {
    coralShooterInterface.setPIDLeft(newkP, newkI, newkD);
  }

  public void setkPLeft(double newkP) {
    coralShooterInterface.setPIDLeft(newkP, values.kILeft, values.kDLeft);
  }

  public void setkILeft(double newkI) {
    coralShooterInterface.setPIDLeft(values.kPLeft, newkI, values.kDLeft);
  }

  public void setkDLeft(double newkD) {
    coralShooterInterface.setPIDLeft(values.kPLeft, values.kILeft, newkD);
  }

  public void setPIDRight(double newkP, double newkI, double newkD) {
    coralShooterInterface.setPIDRight(newkP, newkI, newkD);
  }

  public void setkPRight(double newkP) {
    coralShooterInterface.setPIDRight(newkP, values.kIRight, values.kDRight);
  }

  public void setkIRight(double newkI) {
    coralShooterInterface.setPIDRight(values.kPRight, newkI, values.kDRight);
  }

  public void setkDRight(double newkD) {
    coralShooterInterface.setPIDRight(values.kPRight, values.kIRight, newkD);
  }

  @Override
  public void CS_periodic() {
    coralShooterInterface.updateInputs(values);
  }

  @Override
  public void initDashboard() {
    println("Initializing Dashboard");

    // Using SmartDashboard to tune PIDs
    // --------------------------------------------------
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/Gains/Left/P", CoralShooterConstants.gainsLeft.kP());
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/Gains/Left/I", CoralShooterConstants.gainsLeft.kI());
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/Gains/Left/D", CoralShooterConstants.gainsLeft.kD());
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/Gains/Right/P", CoralShooterConstants.gainsRight.kP());
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/Gains/Right/I", CoralShooterConstants.gainsRight.kI());
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/Gains/Right/D", CoralShooterConstants.gainsRight.kD());

    SmartDashboard.putNumber("Subsystem/CoralShooterSubsystem/LastShotIn(ms)", 0);

    SmartDashboard.putBoolean("Commands/CoralShooterRampUp/OverrideRPM", false);
    SmartDashboard.putNumber("Commands/CoralShooterRampUp/ForcedRMPLeft", 1275);
    SmartDashboard.putNumber("Commands/CoralShooterRampUp/ForcedRMPRight", 1275);
    SmartDashboard.putNumber(
        "Commands/ToPathAndFinkleAndCoralShootWithOffset/OffsetDistance(in)", 7.0);
    SmartDashboard.putNumber("Commands/DriveToPoseFinkle/OffsetDistance(in)", 7.0);

    SmartDashboard.putNumber("Subsystem/PresetManager/AimPreset/rmpErrorMultiplier", 1.0);
    SmartDashboard.putNumber("Subsystem/PresetManager/AimPreset/angleAdjust(deg)", 0);
  }

  @Override
  public void updateDashboard() {
    // Using SmartDashboard to tune PIDs
    // --------------------------------------------------
    double newkPLeft =
        SmartDashboard.getNumber("Subsystem/CoralShooterSubsystem/Gains/Left/P", values.kPLeft);
    double newkILeft =
        SmartDashboard.getNumber("Subsystem/CoralShooterSubsystem/Gains/Left/I", values.kILeft);
    double newkDLeft =
        SmartDashboard.getNumber("Subsystem/CoralShooterSubsystem/Gains/Left/D", values.kDLeft);
    double newkPRight =
        SmartDashboard.getNumber("Subsystem/CoralShooterSubsystem/Gains/Right/P", values.kPRight);
    double newkIRight =
        SmartDashboard.getNumber("Subsystem/CoralShooterSubsystem/Gains/Right/I", values.kIRight);
    double newkDRight =
        SmartDashboard.getNumber("Subsystem/CoralShooterSubsystem/Gains/Right/D", values.kDRight);

    // Coefficients on SmartDashboard have changed, save new values to the PID controller
    // --------------------------------------------------
    values.kPLeft =
        CS_Utils.updateFromSmartDashboard(newkPLeft, values.kPLeft, (value) -> setkPLeft(value));
    values.kILeft =
        CS_Utils.updateFromSmartDashboard(newkILeft, values.kILeft, (value) -> setkILeft(value));
    values.kDLeft =
        CS_Utils.updateFromSmartDashboard(newkDLeft, values.kDLeft, (value) -> setkDLeft(value));
    values.kPRight =
        CS_Utils.updateFromSmartDashboard(newkPRight, values.kPRight, (value) -> setkPRight(value));
    values.kIRight =
        CS_Utils.updateFromSmartDashboard(newkIRight, values.kIRight, (value) -> setkIRight(value));
    values.kDRight =
        CS_Utils.updateFromSmartDashboard(newkDRight, values.kDRight, (value) -> setkDRight(value));

    // Update the SmartDashboard with the current state of the subsystem
    SmartDashboard.putBoolean("Subsystem/CoralShooterSubsystem/Shooter", values.shooterIsEnabled);
    SmartDashboard.putBoolean("Subsystem/CoralShooterSubsystem/Launcher", values.launchIsEnabled);

    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/CurrentRPMLeft", values.currentRPMLeft.in(RPM));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/CurrentRPMRight", values.currentRPMRight.in(RPM));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/CurrentRPMLauncher", values.currentRMPLauncher.in(RPM));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/LauncherSetpoint", values.currentLauncherSetpoint);

    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/ShooterAmpsLeft", values.ampsLeft.in(Amps));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/ShooterAmpsRight", values.ampsRight.in(Amps));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/LauncherAmps", values.ampsLauncher.in(Amps));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/TemperatureShooterLeft", values.tempLeft.in(Celsius));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/TemperatureShooterRight", values.tempRight.in(Celsius));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/TemperatureLauncher", values.tempLauncher.in(Celsius));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/AppliedOutputShooterLeft", values.appliedOutputLeft);
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/AppliedOutputShooterRight", values.appliedOutputRight);
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/AppliedOutputLauncher", values.appliedOutputLauncher);
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/ShooterappliedOutputLeft", values.appliedOutputLeft);
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/ShooterappliedOutputRight", values.appliedOutputRight);
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/LauncherappliedOutputAmps", values.appliedOutputLauncher);

    SmartDashboard.putBoolean("Subsystem/CoralShooterSubsystem/IsLoaded", values.isLoaded);
    SmartDashboard.putBoolean(
        "Subsystem/CoralShooterSubsystem/CoralDetected", values.coralDetected);

    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/ShootingRPMLeft", shootingRPMLeft.in(RPM));
    SmartDashboard.putNumber(
        "Subsystem/CoralShooterSubsystem/ShootingRPMRight", shootingRPMRight.in(RPM));
  }

  // Characterization methods
  public void runCharacterizationLeft(double input) {
    coralShooterInterface.runCharacterizationLeft(input);
  }

  public void runCharacterizationRight(double input) {
    coralShooterInterface.runCharacterizationRight(input);
  }

  public double getCharacterizationVelocityLeft() {
    return values.currentRPMLeft.in(RPM);
  }

  public double getCharacterizationVelocityRight() {
    return values.currentRPMRight.in(RPM);
  }
}

File: src/main/java/frc/robot/subsystems/coralshooter/tsfiles/CoralShooterSubsystem.ts
// CoralShooterSubsystem.ts

import { SmartDashboard } from 'edu.wpi.first.wpilibj.smartdashboard';
import { CS_SubsystemBase } from '../CS_SubsystemBase';
import { CoralShooterInterface, CoralShooterValues } from './CoralShooterInterface';
import { CoralShooterConstants } from './CoralShooterConstants';
import { CS_Utils } from '../utils/CS_Utils';

export class CoralShooterSubsystem extends CS_SubsystemBase {
    private coralShooterInterface: CoralShooterInterface;
    private values: CoralShooterValues = new CoralShooterValues();
    private desiredRPM: number = CoralShooterConstants.shootRPM;
    private desiredLauncherSetpoint: number = CoralShooterConstants.launcherShootSetpoint;

    constructor(subsystemInterface: CoralShooterInterface) {
        super();
        this.coralShooterInterface = subsystemInterface;
        console.log('CoralShooterSubsystem initialized');
    }

    // Interface Calls
    public startRampUp(): void {
        this.coralShooterInterface.startShooter(this.desiredRPM);
    }

    public startShooter(newRPM: number): void {
        this.desiredRPM = newRPM;
        this.coralShooterInterface.startShooter(newRPM);
    }

    public setShooterRPM(newRPM: number): void {
        this.desiredRPM = newRPM;
        this.coralShooterInterface.updateShooterRPM(newRPM);
    }

    public setShootingRPM(): void {
        this.coralShooterInterface.updateShooterRPM(this.desiredRPM);
    }

    public startIntake(): void {
        this.coralShooterInterface.startShooter(CoralShooterConstants.intakeRPM);
        this.coralShooterInterface.startLauncher(CoralShooterConstants.launcherIntakeSetpoint);
    }

    public startLauncher(): void {
        this.coralShooterInterface.startLauncher(this.desiredLauncherSetpoint);
    }

    public startLauncherWithSetpoint(newSetpoint: number): void {
        this.desiredLauncherSetpoint = newSetpoint;
        this.coralShooterInterface.startLauncher(newSetpoint);
    }

    public stopShooter(): void {
        this.coralShooterInterface.stopShooter();
    }

    public stopLauncher(): void {
        this.desiredLauncherSetpoint = 0;
        this.coralShooterInterface.stopLauncher();
    }

    public stopAll(): void {
        this.coralShooterInterface.stopShooter();
        this.coralShooterInterface.stopLauncher();
    }

    public getShooterRPMLeft(): number {
        return this.coralShooterInterface.getShooterRPMLeft();
    }

    public getShooterRPMRight(): number {
        return this.coralShooterInterface.getShooterRPMRight();
    }

    public isLoaded(): boolean {
        return this.coralShooterInterface.shooterIsLoaded();
    }

    public setPID(newkP: number, newkI: number, newkD: number): void {
        this.coralShooterInterface.setPID(newkP, newkI, newkD);
    }

    public setkP(newkP: number): void {
        this.coralShooterInterface.setPID(newkP, this.values.kI, this.values.kD);
    }

    public setkI(newkI: number): void {
        this.coralShooterInterface.setPID(this.values.kP, newkI, this.values.kD);
    }

    public setkD(newkD: number): void {
        this.coralShooterInterface.setPID(this.values.kP, this.values.kI, newkD);
    }

    public CS_periodic(): void {
        this.coralShooterInterface.updateInputs(this.values);
    }

    public initDashboard(): void {
        console.log('Initializing Dashboard');
        SmartDashboard.putNumber('Subsystem/CoralShooter/Gains/P', CoralShooterConstants.gains.kP());
        SmartDashboard.putNumber('Subsystem/CoralShooter/Gains/I', CoralShooterConstants.gains.kI());
        SmartDashboard.putNumber('Subsystem/CoralShooter/Gains/D', CoralShooterConstants.gains.kD());
    }

    public updateDashboard(): void {
        const newkP = SmartDashboard.getNumber('Subsystem/CoralShooter/Gains/P', this.values.kP);
        const newkI = SmartDashboard.getNumber('Subsystem/CoralShooter/Gains/I', this.values.kI);
        const newkD = SmartDashboard.getNumber('Subsystem/CoralShooter/Gains/D', this.values.kD);

        this.values.kP = CS_Utils.updateFromSmartDashboard(newkP, this.values.kP, (value) => this.setkP(value));
        this.values.kI = CS_Utils.updateFromSmartDashboard(newkI, this.values.kI, (value) => this.setkI(value));
        this.values.kD = CS_Utils.updateFromSmartDashboard(newkD, this.values.kD, (value) => this.setkD(value));

        SmartDashboard.putBoolean('Subsystem/CoralShooter/Shooter', this.values.shooterIsEnabled);
        SmartDashboard.putBoolean('Subsystem/CoralShooter/Launcher', this.values.launchIsEnabled);

        SmartDashboard.putNumber('Subsystem/CoralShooter/Shooter RPM Left', this.values.currentRPMLeft);
        SmartDashboard.putNumber('Subsystem/CoralShooter/Shooter RPM Right', this.values.currentRPMRight);
        SmartDashboard.putNumber('Subsystem/CoralShooter/Launcher Setpoint', this.desiredLauncherSetpoint);

        SmartDashboard.putNumber('Subsystem/CoralShooter/Shooter Amps Left', this.values.ampsLeft);
        SmartDashboard.putNumber('Subsystem/CoralShooter/Shooter Amps Right', this.values.ampsRight);
        SmartDashboard.putNumber('Subsystem/CoralShooter/Launcher Amps', this.values.ampsLauncher);

        SmartDashboard.putBoolean('Subsystem/CoralShooter/isLoaded', this.values.isLoaded);

        const newRPM = SmartDashboard.getNumber('Subsystem/CoralShooter/Shooter DesiredRPM', CoralShooterConstants.shootRPM);
        if (newRPM !== this.desiredRPM) {
            this.setShooterRPM(newRPM);
        }
        SmartDashboard.putNumber('Subsystem/CoralShooter/Shooter DesiredRPM', this.desiredRPM);
    }

    public runCharacterization(input: number): void {
        this.coralShooterInterface.runCharacterizationLeft(input);
        this.coralShooterInterface.runCharacterizationRight(input);
    }

    public getCharacterizationVelocity(): number {
        return (this.values.currentRPMLeft + this.values.currentRPMRight) / 2.0;
    }
}